test-basic/concurrent-requests.js

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2020 MarkLogic Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var testconfig = require('../etc/test-config.js');
+
+var marklogic = require('../');
+
+var dbWriter = marklogic.createDatabaseClient(testconfig.restWriterConnection);
+
+describe('concurrent-requests test', function(){
+
+
+    it('should write 1000 documents concurrently', function(done){
+        for(var i=0; i<3; i++) {
+            writeBatch(i,0);
+        }
+
+        done();
+    });
+});
+
+
+function writeBatch(writerId, nextDocument) {
+    if(nextDocument>=1000) {
+        console.log(writerId + ' is done');
+        return;
+    }
+    var documentDescriptors = createDocumentDescriptors();
+    console.log('writerId is '+writerId);
+    console.log('writing batch '+nextDocument+' to '+(nextDocument+10));
+    var batch = documentDescriptors.slice(nextDocument, nextDocument +10);
+    nextDocument = nextDocument+10;
+
+    dbWriter.documents.write(batch).result(writeBatch(writerId, nextDocument)).catch();
```

Comment:
Another part of the issue might be here.  Line 47 calls writeBatch() immediately and passes the return value to result.

In other words, the calls to writeBatch are recursing here such that the 1st iteration of the for loop processes all of the items in the array.  There's nothing left for the other 2 iterations of the loop.

Instead, line 47 should pass a callback that executes when the write() request finishes.  That's what allows other requests to execute.

Creating and passing a lambda function should work:

    .result((output) => writeBatch(writerId, nextDocument, done))

Also, .catch() should at a minimum provide a logging callback such as

    .catch((err) => console.log(err));

Otherwise, the error is just silently ignored -- similar to a Java catch clause that does nothing with the caught exception.